### PR TITLE
feat: Optimize StaticHtml component for React js 🚀 

### PR DESCRIPTION
**Pull request by sanjaiyan-dev**
This PR optimizes the StaticHtml component, addressing the limitations of the previous implementation (which relied on an incorrect `shouldComponentUpdate` strategy).  It now correctly uses `React.memo` to prevent unnecessary re-renders of the static HTML content.  This approach is more performant and is the correct way to memoize function components.

*   **Fixes**: Incorrect `shouldComponentUpdate` implementation.
*   **Impact**: Improves the performance of rendering static HTML content in React.
*   **Why**: Makes use of modern react patterns, improves performance and fixes errors.
*   **Implementation Details**: Replaced `shouldComponentUpdate` with `React.memo` using custom comparison function to prevent re-renders.

https://react.dev/reference/react/memo

### CODE REVIEW

File: packages/integrations/react/src/static-html.ts [hunk 1]
```diff
@@ -26,12 +26,10 @@ const StaticHtml = ({
...
+ * React.memo is the modern functional equivalent of shouldComponentUpdate.
+ * 
+ * By returning `true` in the comparison function (the second argument),
+ * we tell React that the props are "equal" and it should skip re-rendering,
+ * effectively making this subtree static.
  */
-StaticHtml.shouldComponentUpdate = () => false;
-
-export default StaticHtml;
+export default memo(StaticHtml, () => true);
```
Comment (florian-lefebvre) [inline packages/integrations/react/src/static-html.ts:35]: Just noting that `memo` is indeed noted as the equivalent at the end of https://react.dev/reference/react/Component#shouldcomponentupdate. Memo docs: https://react.dev/reference/react/memo

Comment (sanjaiyan-dev) [reply packages/integrations/react/src/static-html.ts:35]: > Just noting that `memo` is indeed noted as the equivalent at the end of https://react.dev/reference/react/Component#shouldcomponentupdate. Memo docs: https://react.dev/reference/react/memo

Thanks for checking! You are correct that they serve the same purpose. But actually, `shouldComponentUpdate` is only for Class components, so it gets ignored here. The old code wasn't really running. Switching to `memo` ensures it works properly.

Sorry if I am wrong

Comment (florian-lefebvre) [reply packages/integrations/react/src/static-html.ts:35]: I wouldn't be surprised if it worked, undocumented

File: .changeset/tough-days-sell.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@astrojs/react': major
+---
+
+Fix StaticHtml performance by implementing `React.memo` for proper re-render prevention.
```
Comment (florian-lefebvre) [inline .changeset/tough-days-sell.md]:
```suggestion
---
'@astrojs/react': patch
---

Improves rendering of static HTML
```

Comment (sanjaiyan-dev) [reply .changeset/tough-days-sell.md]: Hi @florian-lefebvre, is this PR correct? Sorry for any inconvenience.

**Comment by sanjaiyan-dev** [comment]
Hi @florian-lefebvre, sorry to bother you again. Could you please let me know if this PR looks good now ?

File: .changeset/tough-days-sell.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@astrojs/react': patch
+---
+
+Improves rendering of static HTML
```
Comment (sarah11918) [inline .changeset/tough-days-sell.md]:
```suggestion
Refactors to improve the performance of rendering static HTML content in React
```
nit: it's not the "rendering" itself that's improved, it's the *performance*, right?

A slightly more precise statement helps people understand whether or not they need to care/worry about the change, and whether they might need to update their own code. So a word like "refactors" is handy here, because it implies that nothing should have really changed for *their* project!

Alternatively, "Improves the performance of rendering static HTML content in React" is perfectly fine, too! But something that emphasizes that it's a *performance* improvement (only) I think is most helpful here as a message to users!

**Comment by sanjaiyan-dev** [comment]
> Thanks for making Astro better, @sanjaiyan-dev ! 🚀 Quick note from docs on the changeset!

Thanks 🙏

**Review by sarah11918 — APPROVED** [review]
Approving for docs! 🥳

**Comment by sanjaiyan-dev** [comment]
Hi @florian-lefebvre ,

I tested two scenarios:
1.  **Micro-interactions (100 updates):** Simulating a user typing quickly into an input field.
2.  **Stress Test (100,000 updates):** Simulating heavy, continuous state changes to measure raw CPU overhead.

### 📊 Benchmark Results

| Scenario | Old Implementation (Unoptimized) | **New Implementation (PR #14917)** | Improvement |
| :--- | :--- | :--- | :--- |
| **100 Updates** (UI Interaction) | ~32.08ms | **~12.17ms** | **~62% Faster** ⚡ |
| **100k Updates** (Stress Test) | ~3658.26ms | **~3030.30ms** | **~17% Faster** |

*(Benchmarks run on Node.js with `react-test-renderer`)*

### 💡 Key Takeaways
The **62% improvement** on the smaller dataset is the real winner here. In real-world usage (like a search bar or a toggle next to heavy static content), this change significantly reduces the "cost per frame," helping to prevent input lag. By switching to `React.memo`, we are successfully short-circuiting the diffing process for the static HTML entirely.

<details>
<summary><b>🛠️ Click to view the benchmark code (benchmark.mjs)</b></summary>

```javascript
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { performance } from 'perf_hooks';

// 1. Old Implementation (Buggy, Pre-PR #14917)
const OldStaticHtml = ({ value, name }) => {
  if (!value) return null;
  return React.createElement('astro-slot', {
    name,
    dangerouslySetInnerHTML: { __html: value },
  });
};
OldStaticHtml.shouldComponentUpdate = () => false; // Ignored by React!

// 2. New Implementation (Optimized via PR #14917)
const NewStaticHtml = React.memo(({ value, name }) => {
  if (!value) return null;
  return React.createElement('astro-slot', {
    name,
    dangerouslySetInnerHTML: { __html: value },
  });
}, () => true); // Forces React to skip re-renders

// 3. Parent Component to simulate updates
function App({ count, ComponentToTest }) {
  // Simulate a massive block of static HTML generated by Astro
  const heavyHtml = "<div class='heavy'>".repeat(100) + "Astro Content" + "</div>".repeat(100);
  
  return React.createElement('div', null, 
    React.createElement('h1', null, `Render Count: ${count}`),
    React.createElement(ComponentToTest, { value: heavyHtml, name: "default" })
  );
}

// 4. Benchmark Runner
function runBenchmark(name, ComponentToTest, iterations = 100_000) {
  // Initial render
  const root = TestRenderer.create(React.createElement(App, { count: 0, ComponentToTest }));
  
  const start = performance.now();
  
  // Trigger continuous re-renders (simulating rapid state changes)
  for (let i = 1; i <= iterations; i++) {
    root.update(React.createElement(App, { count: i, ComponentToTest }));
  }
  
  const end = performance.now();
  console.log(`${name}: ${(end - start).toFixed(2)}ms for ${iterations} updates`);
}

console.log('🚀 Benchmarking StaticHtml React Integration...\n');
runBenchmark('❌ Old StaticHtml (Unoptimized)', OldStaticHtml);
runBenchmark('✅ New StaticHtml (Optimized)', NewStaticHtml);
```
</details>

**Comment by sanjaiyan-dev** [comment]
<img width="1046" height="802" alt="<StaticHTML> benchmark before" src="https://github.com/user-attachments/assets/6d2b0e83-a6e8-4884-899e-198c98d682b0" />
<img width="1032" height="690" alt="<StaticHTML> benchmark after" src="https://github.com/user-attachments/assets/27ff581c-a514-499b-9572-dd5e226dca0b" />

File: packages/integrations/react/src/static-html.ts (hunk 1 above)
Comment (sanjaiyan-dev) [reply packages/integrations/react/src/static-html.ts:35]: > I wouldn't be surprised if it worked, undocumented

It appears the previous implementation was ineffective and had no measurable impact. Here is the evidence:

Screenshot 2026-03-07 at 14 48 18


<details>
<summary><strong>View Benchmark Code</strong></summary>

```javascript
import React, { useState } from 'react';
import TestRenderer from 'react-test-renderer';
import { performance } from 'perf_hooks';

// --- 1. Setup: Render Counters ---
let oldRenderCount = 0;
let newRenderCount = 0;

// --- 2. Components ---
const OldStaticHtml = ({ value, name }) => {
	oldRenderCount++; // Track how many times React actually runs this function
	return React.createElement('astro-slot', { name, dangerouslySetInnerHTML: { __html: value } });
};
OldStaticHtml.shouldComponentUpdate = () => false;

const NewStaticHtml = React.memo(({ value, name }) => {
	newRenderCount++; // Track how many times React actually runs this function
	return React.createElement('astro-slot', { name, dangerouslySetInnerHTML: { __html: value } });
});

// --- 3. The "Real World" Parent ---
// This component simulates a page that re-renders frequently (e.g., a timer)
function App({ ComponentToTest, value }) {
	return React.createElement(
		'div',
		{ className: 'page' },
		React.createElement('h1', null, 'App Header'),
		React.createElement(ComponentToTest, { value, name: 'content' }),
	);
}

// --- 4. The Benchmark Runner ---
function runBenchmark(label, ComponentToTest, iterations = 10000) {
	const htmlContent = '<div>Static Content</div>';
	const root = TestRenderer.create(
		React.createElement(App, { ComponentToTest, value: htmlContent }),
	);

	// Reset counters
	oldRenderCount = 0;
	newRenderCount = 0;

	const start = performance.now();

	// Simulate 1000 parent re-renders
	for (let i = 0; i < iterations; i++) {
		root.update(React.createElement(App, { ComponentToTest, value: htmlContent }));
	}

	const end = performance.now();
	const count = label.includes('Old') ? oldRenderCount : newRenderCount;

	console.log(`${label}:`);
	console.log(` - Time: ${(end - start).toFixed(2)}ms`);
	console.log(` - Actual Render Calls: ${count}`);
	console.log('-----------------------------------');
}

console.log('🚀 Running Real-World Simulation (10000 Parent Updates)...\n');
runBenchmark('❌ Old (No Memo)', OldStaticHtml);
runBenchmark('✅ New (Memo)', NewStaticHtml);
```
</details>

Comment (sanjaiyan-dev) [reply packages/integrations/react/src/static-html.ts:35]: I ran some benchmarks to test the impact of memoization on components with heavy computations (simulating real-world scenarios like expensive data processing or complex logic).

The results show that in complex cases, the performance gain is massive—nearly **205x faster** compared to unmemoized components 🚀

Astro js with react js benchmark(memo vs  unoptimised) screenshot


<details>
<summary>Click to see the Benchmark Code</summary>

```tsx
// @vitest-environment jsdom
import { bench, describe } from 'vitest';
import React, { memo } from 'react';
import { render } from '@testing-library/react';

// Simulates heavy work (e.g., complex integration/data processing)
function findNthPrime(n: number): number {
  let count = 0;
  let num = 2;
  while (count < n) {
    let isPrime = true;
    for (let i = 2; i <= Math.sqrt(num); i++) {
      if (num % i === 0) {
        isPrime = false;
        break;
      }
    }
    if (isPrime) count++;
    num++;
  }
  return num - 1;
}

const LIST_SIZE = 300;
const STATIC_ITEMS = Array.from({ length: LIST_SIZE / 2 }, (_, i) => `Static-${i}`);
const DYNAMIC_ITEMS = Array.from({ length: LIST_SIZE / 2 }, (_, i) => i);

const DynamicItem = ({ index, searchText }: { index: number; searchText: string }) => {
  const complexValue = findNthPrime(300 + (index % 50)).toExponential(); 
  return (
    <div>
      <strong>Item {index}:</strong> {searchText} 
      <small>(Calculated: {complexValue})</small>
    </div>
  );
};

const ListComponent = ({ searchText }: { searchText: string }) => (
  <div>
    {STATIC_ITEMS.map((item) => <div key={item}>Fixed: {item}</div>)}
    {DYNAMIC_ITEMS.map((i) => <DynamicItem key={i} index={i} searchText={searchText} />)}
  </div>
);

const MemoizedListComponent = memo(ListComponent, () => true);

const App = ({ text, memoEnabled }: { text: string; memoEnabled: boolean }) => (
  <div>
    {memoEnabled ? <MemoizedListComponent searchText={text} /> : <ListComponent searchText={text} />}
  </div>
);

const { rerender: rerenderNormal } = render(<App text="init" memoEnabled={false} />);
const { rerender: rerenderMemo } = render(<App text="init" memoEnabled={true} />);

describe('Complex React Performance', () => {
  bench('Normal (Re-calculates on every type)', () => {
    rerenderNormal(<App text={Math.random().toString()} memoEnabled={false} />);
  });

  bench('Memoized (Skips Calculation)', () => {
    rerenderMemo(<App text={Math.random().toString()} memoEnabled={true} />);
  });
});
```
</details>

This confirms that while simple renders might not show a huge difference, when heavy work is involved, preventing unnecessary re-renders is critical. This is exactly why deep dependency tracking (like what the React Compiler aims to do) is so valuable for complex UIs.

Anyway, I wanted to share these findings—sorry if I'm mistaken, but I believe this demonstrates the real-world value of these optimizations!

**Comment by sanjaiyan-dev** [comment]
Hi @florian-lefebvre  and @matthewp  👋

I totally understand the need to be cautious here since React hydration is so sensitive.

Are there any specific edge cases (like nested islands or dynamic slots) you're worried about? I'd be happy to write some extra tests for those to prove it's safe and help you feel confident moving this forward.

Just let me know how I can help :)

**Comment by sanjaiyan-dev** [comment]
> Researched this a bit, asked AI etc and it seems fine.

Thanks :)
